### ISSUE TITLE
Question about Pipeline parallelism

### ISSUE THREAD

**Issue by vermouth1992** [pytorch/torchtitan#431]
Just wonder does the current PipelineStage API supports variable length input shapes like in Megatron? https://github.com/NVIDIA/Megatron-LM/blob/e33c8f78a35765d5aa37475a144da60e8a2349d1/megatron/core/model_parallel_config.py#L212 This is particular useful for packed inputs where all the paddings are removed.

**Comment by H-Huang** [comment]
The `PipelineStage` API currently does not support variable length input shapes. We have an open issue which tracks a similar feature (lazy shape inference) that which could potentially be used to support variable length inputs (https://github.com/pytorch/pytorch/issues/126216).

Wanted to understand more about the use case though. Can you elaborate on `This is particular useful for packed inputs where all the paddings are removed`, is there an example?

**Comment by vermouth1992** [comment]
The current models implemented in huggingface transformers library use input_ids and attention_mask as inputs. For SFT and RLHF, a large portion of the inputs will be padding. Computing those paddings through MLP and attention is a huge waste. Instead, we can unpad the data (e.g. https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/bert_padding.py#L98) so that all the sequences inside a batch is packed and can separated via cu_seqlens.
For MLP and layernorm, we don't need to change the model as there is no inter-token computation. For attention, flash-attention supports passing packed inputs. https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_attn_interface.py#L1005
In this case, every micro-batch will have a different input shapes during each forward. Currently, this can be workaround by passing padded input_ids and attention_masks and call unpad inside the model of each pipeline stage, with some loss of performance.

**Comment by wconstab** [comment]
The root issue here is that a recv operation needs to know its size before starting. So every stage besides 0 needs to know the size of the new microbatch before passing the data down the pipeline.

A couple ways this can be done-
1. We require every pipeline stage to receive data from the data loader locally. This puts extra load on the data loader but may be ok. Then we need some way to infer the input shape of the current stage as a function of the microbatch shape. Then we can directly allocate recv buffers for each batch.
2. We add an extra communication step into our schedules that passes the output shape from our forward to the next stage, which then allocates a buffer and performs recv. This extra communication will add overhead not only due to nccl time, but maybe also additional bubble time if it prevents us starting the recv operation as early.

In the current situation, you pay extra communication cost for padding, but then you unpad the data before using it. Is that correct? We'd have to understand how bad this overhead is, compared to the above proposals and their overhead and complexity. Do you have some data?

**Comment by kwen2501** [comment]
The Megatron code has the following comment:
> Setting this communicates the size of tensors during pipeline parallelism communication, because of this extra overhead it should only be set if the sequence length varies by microbatch within a global batch.

My understanding of the above comment is that the microbatch size is still constant , but the number of microbatches could change.

Does that apply to your case? 

If it applies, it should be doable by reconfiguring the num_microbatches of PipelineSchedule dynamically. (We don't provide such an API now but it should be doable.)

The other challenge is making such reconfiguration on all pipeline ranks because they need to agree on it -- do you think you could make such API call on all ranks? Otherwise, as the Megatron comment suggests, there will be an overhead for rank 0 to inform all other ranks.

Cc @wconstab @H-Huang

**Comment by vermouth1992** [comment]
Actually, the microbatch size varies inside a batch across microbatches. In Megatron, the size of the output of each TransformerLayer is [total_nnz, hidden_size], which total_nnz is the total number of tokens inside the microbatch after removing all the paddings.

Assume we have a batch of size [batch_size, seqlen, hidden_size]. We truncate in the batch dimension into a list of microbatches [batch_size // n_microbatches, seqlen, hidden_size]. Then, we perform unpad operations, and it becomes [total_nnz, hidden_size], which total_nnz is the total number of tokens inside the microbatch after removing all the paddings.

> Support for variable sequence lengths across microbatches. Setting this communicates the size of tensors during pipeline parallelism communication, because of this extra overhead it should only be set if the sequence length varies by microbatch within a global batch.

**Comment by man2machine** [comment]
I wanted to just ping this issue again. Variable microbatch sizes for pipeline parallelism definitely is a useful feature to add. In some use cases I was working on, I ran into a scenario where this was needed. Similar to Megatron, there can be a flag to set which enables the extra overhead of communicating the tensor shapes, but allows variable microbatch sizes.

**Comment by wconstab** [comment]
I think this is a reasonable ask, we never prioritized working on it, and I also worried about the code complexity.  I'd be happy to take a PR to add this if it's clean.  cc @H-Huang @sanketpurandare

**Comment by acisseJZhong** [comment]
this feature is also useful for verl integration work! For context, I am trying to add torch distributed PP to Verl and faced similar issue https://github.com/pytorch/torchtitan/issues/2370#issuecomment-3886565207, another option is to pad to max seq_len across microbatches, and pass in to the pp_schedule.

**Comment by wconstab** [comment]
well, i still want to ask: do you _really_ want varlen pp in verl? it will add overhead.  will this be acceptable, or would you be forced to go the padding route to get performance back?

**Comment by acisseJZhong** [comment]
good question! I think padding microbatches should also work relatively well compared to the naive case(as we already pack multiple samples together into batch size 1 and remove extra paddings). So I should probably implement this first. 

@vermouth1992 let me know if you have different opinions.

**Comment by vermouth1992** [comment]
> well, i still want to ask: do you _really_ want varlen pp in verl? it will add overhead. will this be acceptable, or would you be forced to go the padding route to get performance back?

What is the overhead?

**Comment by vermouth1992** [comment]
To support varlen pp, we can either precompute the shape and pass to the pp schedule, or let the pp schedule communicate shapes during runtime. The overhead is negligible in both cases

**Comment by wconstab** [comment]
> what is the overhead?

my concern about overhead was coming from the need to move data synchronously between gpu and cpu at the send/recv points in the PP schedule, hurting the ability of the schedule to overlap comms and compute. I did not actually test this or measure the amount of overhead.  The mechanism for the overhead would be that you need to know the recv buffer size on the CPU before you issue the recv op for the varlen tensor data.  the recv buffer size would be rec'd as a GPU tensor, and have to be moved to CPU.  I think the net result of this would be that every PP comm goes from one 'send-recv' pair to a sequence of 
1) move shape data from CPU to GPU  <-- sync point
2) send/recv shape on gpu
3) move shape back to CPU <-- sync point
4) send/recv the tensordata on gpu

I'm curious- have you benchmarked this, I know megatron supports varlen PP somehow, but i haven't taken a look at a profile trace to see how expensive this is, or if there is a smarter way to handle the cpu/gpu sync?


> precompute the shape and pass to the pp schedule,

well, this approach would work. it's a bit invasive to the APIs, but, i would be open to it. it might be better than the automatic approach due to the overhead.
